api-server/ui/Map.tsx

```diff
@@ -1,27 +1,46 @@
+import { SKContext } from '@chacal/signalk-ts'
 import * as React from 'karet'
 import * as U from 'karet.util'
+import { Observable } from 'kefir'
+import { Atom } from 'kefir.atom'
 import { LatLngBounds, LeafletEvent } from 'leaflet'
+import * as L from 'partial.lenses'
 import { GeoJSON, Map as LeafletMap, TileLayer } from 'react-leaflet'
 import { Coords, TrackGeoJSON } from '../domain/Geo'
-import { Atomized } from './ui-domain'
+import updateTracksFor from './trackprovider'
+import { LoadState, mapArrayInObs, Vessel } from './ui-domain'
 
 // Lift LeafletMap to Karet to support Atoms as props
 const KaretMap = U.toKaret(LeafletMap)
 
+export interface Track {
+  context: SKContext
+  geoJson: TrackGeoJSON
+}
+
 export interface MapProps {
-  center: Coords
-  zoom: number
-  bounds: LatLngBounds
-  tracks: TrackGeoJSON
+  center: Observable<Coords, any>
+  zoom: Atom<number>
+  bounds: Atom<LatLngBounds>
+  vessels: Atom<Vessel[]>
 }
 
-const Map = ({ center, zoom, bounds, tracks }: Atomized<MapProps>) => {
+const Map = ({ center, zoom, bounds, vessels }: MapProps) => {
   const updateBoundsFromMap = (comp: LeafletMap) =>
     bounds.set(comp.leafletElement.getBounds())
   const updateBoundsFromEvent = (e: LeafletEvent) => {
     bounds.set(e.target.getBounds())
     zoom.set(e.target.getZoom())
   }
+  const vesselsWithTracks = vessels.map(vessels =>
+    vessels.filter(vesselHasTracks)
+  )
+
+  // Update tracks in global state when vessels change
+  updateTracksFor(vessels, zoom, bounds)
```

Comment:
Not sure what the `updateTracksFor` function does, but generally speaking you should avoid directly executing side-effects in the body of a component function.

One technical problem is that it may force components to be updated at a point where React is processing them.  React will then give you error/warning messages and things wont work correctly.

More fundamentally components should be essentially pure functions from state to VDOM.  You want components to be update as a side effect of state modification and not the other way around (having component update cause state modification as a side effect).

Reply:
`updateTracksFor` subscribes to `zoom` & `bounds` and reloads tracks when they change. So it's only state management and no rendering whatsoever.

What would be the most suitable place for initializing such state modifying streams? Maybe in the `index.tsx` after `ReactDOM.render`? (After reading your comment, I reckon that those probably should not be inside any component, right?)

PS: Huge thanks for checking out the code and giving feedback! Highly appreciated!